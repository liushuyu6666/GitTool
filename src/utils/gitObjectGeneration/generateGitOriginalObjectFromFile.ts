import fs from 'fs';
import path from 'path';
import { inflateSync } from 'zlib';
import { GitObject } from '../../object/gitObject/GitObject';
import getGitObjectType from '../getGitObjectType';
import splitHeaderAndBody from '../splitHeaderAndBody';

export default function (filePath: string): GitObject {
  const filePaths: string[] = filePath.split(path.sep);
  const suffix = filePaths.pop();
  const prefix = filePaths.pop();
  if (!suffix || !prefix) {
    throw new Error(
      `GitOriginalObjectGenerator: can\'t get prefix or suffix from the file path ${filePath}`,
    );
  }
  const hash = prefix + suffix;

  const decryptedBuf = inflateSync(fs.readFileSync(filePath));
  const [header, _] = splitHeaderAndBody(decryptedBuf);
  const headerStr = header.toString();
  const size = parseInt(headerStr.split(' ')[1]);
  const type = getGitObjectType(headerStr.split(' ')[0]);

  const bodyOffsetStartIndex = header.length + 1;
  const bodyOffsetEndIndex = decryptedBuf.length;

  return new GitObject({
    hash,
    type,
    size,
    filePath: filePath,
    bodyOffsetStartIndex,
    bodyOffsetEndIndex,
  });
}
