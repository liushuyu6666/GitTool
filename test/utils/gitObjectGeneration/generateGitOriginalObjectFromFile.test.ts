import generateGitOriginalObjectFromFile from '../../../src/utils/gitObjectGeneration/generateGitOriginalObjectFromFile';
import { gitOriginalObjectBrief, TestObjectKit } from '../../data/data';

describe('Test generateGitOriginalObjectFromFile, which generates brief object from file. Here three type need to be tested:', () => {
  const testDataKit: TestObjectKit = gitOriginalObjectBrief();
  
  test('Test the blob object', () => {
    const filePath = testDataKit.blob.filePath;
    const gitObject = generateGitOriginalObjectFromFile(filePath);

    expect(gitObject).toMatchSnapshot();
  });

  test('Test the tree object', () => {
    const filePath = testDataKit.tree.filePath;
    const gitObject = generateGitOriginalObjectFromFile(filePath);

    expect(gitObject).toMatchSnapshot();
  });

  test('Test the commit object', () => {
    const filePath = testDataKit.commit.filePath;
    const gitObject = generateGitOriginalObjectFromFile(filePath);

    expect(gitObject).toMatchSnapshot();
  });
});
