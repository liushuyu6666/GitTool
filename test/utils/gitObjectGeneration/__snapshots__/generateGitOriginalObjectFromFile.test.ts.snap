// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test generateGitOriginalObjectFromFile, which generates brief object from file. Here three type need to be tested: Test the blob object 1`] = `
GitObject {
  "bodyOffsetEndIndex": 59,
  "bodyOffsetStartIndex": 8,
  "data": undefined,
  "filePath": "gitTestSimple/.git/objects/21/29cac3bcef43797f755ba6105c32d4a3f1ace0",
  "hash": "2129cac3bcef43797f755ba6105c32d4a3f1ace0",
  "prefix": "21",
  "size": 51,
  "suffix": "29cac3bcef43797f755ba6105c32d4a3f1ace0",
  "type": "blob",
}
`;

exports[`Test generateGitOriginalObjectFromFile, which generates brief object from file. Here three type need to be tested: Test the commit object 1`] = `
GitObject {
  "bodyOffsetEndIndex": 253,
  "bodyOffsetStartIndex": 11,
  "data": undefined,
  "filePath": "gitTestSimple/.git/objects/7c/556ca93b467f8f8247acf522915f2f9e048eb5",
  "hash": "7c556ca93b467f8f8247acf522915f2f9e048eb5",
  "prefix": "7c",
  "size": 242,
  "suffix": "556ca93b467f8f8247acf522915f2f9e048eb5",
  "type": "commit",
}
`;

exports[`Test generateGitOriginalObjectFromFile, which generates brief object from file. Here three type need to be tested: Test the tree object 1`] = `
GitObject {
  "bodyOffsetEndIndex": 43,
  "bodyOffsetStartIndex": 8,
  "data": undefined,
  "filePath": "gitTestSimple/.git/objects/32/6ae25c4e61cba56782e9e37a19d0bbcf9b694e",
  "hash": "326ae25c4e61cba56782e9e37a19d0bbcf9b694e",
  "prefix": "32",
  "size": 35,
  "suffix": "6ae25c4e61cba56782e9e37a19d0bbcf9b694e",
  "type": "tree",
}
`;
